File: src/21/day21.py
def make_steps(index: int, x_bound: int, y_bound: int) -> set[int]:
    steps = set()
    if index // x_bound > 0:
        steps.add(index - x_bound)
    if index // x_bound < y_bound - 1:
        steps.add(index + x_bound)
    if index % x_bound > 0:
        steps.add(index - 1)
    if index % x_bound < x_bound - 1:
        steps.add(index + 1)
    return steps

def part1(filename: str) -> int:
    with open(filename, 'rt') as f:
        text = [x.strip() for x in f.readlines()]
    x_bound = len(text[0])
    y_bound = len(text)
    gardens = set()
    exclusions = set()
    start = 0
    i = 0
    for line in text:
        for character in line:
            match character:
                case '.':
                    gardens.add(i)
                case '#':
                    exclusions.add(i)
                case 'S':
                    start = i
                    gardens.add(i)
            i += 1
    steps = {start}
    for _ in range(64):
        new_steps = set()
        for step in steps:
            new_steps |= make_steps(step, x_bound, y_bound)
        steps = new_steps & gardens
    return len(steps)

def step_outward(index: int, x_bound: int, y_bound: int, start: int) -> set[int]:
    steps = set()
    if 0 < index // x_bound <= start // x_bound:
        steps.add(index - x_bound)
    if start // x_bound <= index // x_bound < y_bound - 1:
        steps.add(index + x_bound)
    if 0 < index % x_bound <= start % x_bound:
        steps.add(index - 1)
    if start % x_bound <= index % x_bound < x_bound - 1:
        steps.add(index + 1)
    return steps

def part2(filename: str) -> int:
    with open(filename, 'rt') as f:
        text = [x.strip() for x in f.readlines()]
    x_bound = len(text[0])
    y_bound = len(text)
    gardens = set()
    exclusions = set()
    start = 0
    i = 0
    for line in text:
        for character in line:
            match character:
                case '.':
                    gardens.add(i)
                case '#':
                    exclusions.add(i)
                case 'S':
                    start = i
                    gardens.add(i)
            i += 1
    explored = {start}
    grid = [0] * (x_bound * y_bound)
    steps = make_steps(start, x_bound, y_bound) & gardens
    for step in steps:
        grid[step] = 1
        explored.add(step)
    step_count = 2
    while steps:
        new_steps = set()
        for step in steps:
            new_steps |= make_steps(step, x_bound, y_bound)
        new_steps &= gardens
        for step in new_steps:
            grid[step] = step_count
            explored.add(step)
        new_steps.difference_update(explored)
        step_count += 1
        steps = new_steps
        print(steps)
    print(grid)
    return 0

if __name__ == '__main__':
    print(part1('day21_input.txt'))
    print(part2('test_input.txt'))

File: src/21/day21.rs
use std::collections::BTreeSet;

fn make_steps(index: usize, x_bound: usize, y_bound: usize) -> BTreeSet<usize> {
    let mut steps = BTreeSet::<usize>::new();
    if index / x_bound > 0 { steps.insert(index - x_bound); }
    if index / x_bound < y_bound - 1 { steps.insert(index + x_bound); }
    if index % x_bound > 0 { steps.insert(index - 1); }
    if index % x_bound < x_bound - 1 { steps.insert(index + 1); }
    return steps;
}

fn part1(filename: &str) -> usize {
    let file_lines = std::fs::read_to_string(filename).expect("File not found");
    let lines = file_lines.trim().split('\n').map(|x| x.trim()).collect::<Vec<_>>();
    let x_bound = lines[0].len();
    let y_bound = lines.len();
    let mut gardens = BTreeSet::<usize>::new();
    let mut start: usize = 0;
    let mut i: usize = 0;
    for line in lines {
        for c in line.chars() {
            match c {
                '.' => {
                    gardens.insert(i);
                },
                'S' => {
                    start = i;
                    gardens.insert(i);
                }
                _ => {}
            }
            i += 1;
        }
    }
    let mut steps = BTreeSet::<usize>::new();
    steps.insert(start);
    for _ in 0..64 {
        let mut new_steps = BTreeSet::<usize>::new();
        for step in steps {
            for elem in make_steps(step, x_bound, y_bound) {
                new_steps.insert(elem);
            }
        }
        steps = new_steps.intersection(&gardens).cloned().collect();
    }
    return steps.len();
}

fn main() {
    println!("{}", part1("day21_input.txt"));
}
